### ISSUE TITLE
Add workarounds for Intel UHD 620/630 driver bug on Windows

### ISSUE THREAD

**Issue by SonoSooS** [libsdl-org/SDL#5628]
As of writing this issue, Intel still hasn't fixed this issue in driver version 30.0.101.1660, and considering that this bug has been around for almost half a decade, with countless issue reports, I don't see them ever fixing this, so community workarounds are needed.

While this bug can manifest itself in many ways, including degraded performance, stutters even though the GPU is rendering more frames than the display refresh rate, and sometimes even complete program or even driver lockups, it's most commonly the first two which happen in practice.

I'm still unable to completely pinpoint how and where this bug actually occurs, but with lots of experiments I do know how to work around this issue. While this issue has only been reproducible on Intel UHD 620 and UHD 630 GPUs, HD 620 and HD 630 might also be affected, but I was unable to find anyone who has these, so I was unable send anyone my test program.

Making this issue even worse, if there is a dedicated GPU along with the iGPU in an Nvidia Optimus setting, the dedicated GPU's output is also hindered by this bug, as the image data is piped through the iGPU, and also displayed by it.
If the integrated GPU is not used, by switching to a dedicated screen controlled by the dedicated GPU, this bug doesn't happen at all.
Switching to the generic Microsoft Display Adapter driver on the integrated GPU just causes stutter in every test setup, even though the dedicated GPU is rendering waaaay more frames than the display's refresh rate.

Edit: forgot the mention, that this only occurs with OpenGL, and with SwapInterval(0). While SwapInterval(1) does not cause frame drops, it does have 1-2frames of latency, which is eliminated by this workaround as well.

The change of code needed to work around this bug is absolutely minimal. Having tested this workaround myself, [this](https://github.com/libsdl-org/SDL/blob/main/src/video/windows/SDL_windowsopengl.c#L855) has to be altered like so:
from
```C
SwapBuffers(hdc);
```
to
```C
HWND hwnd = ((SDL_WindowData *)window->driverdata)->hwnd;

BOOL swapSuccess;

// The workaround should not be used for exclusive fullscreen!
// Fake fullscreen however should still use the workaround.
if((window->flags & SDL_WINDOW_FULLSCREEN_DESKTOP) == SDL_WINDOW_FULLSCREEN)
{
    swapSuccess = SwapBuffers(hdc);
}
else
{
    InvalidateRect(hwnd, 0, 0);
    swapSuccess = wglSwapLayerBuffers(hdc, WGL_SWAP_MAIN_PLANE); // Calls SwapBuffers internally, but SwapBuffers alone doesn't work here for this workaround.
}
```

There is just one downside to just applying this patch directly: calling InvalidateRect (especially in exclusive fullscreen mode, where this issue does not happen) would degrade the performance for everyone else not affected by this bug.
My suggestion in this case would be to detect if the GL vendor string contains UHD 620 or UHD 630, and enable workarounds. And perhaps add in an SDL hint to force this workaround for any GPU, not just the affected ones.

**Comment by Susko3** [comment]
`window->flags & SDL_WINDOW_FULLSCREEN` should be
```cs
(window->flags & SDL_WINDOW_FULLSCREEN_DESKTOP) == SDL_WINDOW_FULLSCREEN
```

to match only exclusive fullscreen.

The former will match [`SDL_WINDOW_FULLSCREEN_DESKTOP`](https://github.com/libsdl-org/SDL/blob/6d1c3cd6918aa84fe2a0ba7d895169a4f69233bc/include/SDL_video.h#L115).

**Comment by SonoSooS** [comment]
Oops, that's what I get for forgetting to research the flags. Good catch! I edited the code block to reflect that.

**Comment by cgutman** [comment]
> As of writing this issue, Intel still hasn't fixed this issue in driver version 30.0.101.1660, and considering that this bug has been around for almost half a decade, with countless issue reports, I don't see them ever fixing this, so community workarounds are needed.

Can you describe in more detail what the bug actually is and how to determine if it's affecting a game? If you have links to the discussion of this issue elsewhere, that would be helpful for context.

> While this bug can manifest itself in many ways, including degraded performance, stutters even though the GPU is rendering more frames than the display refresh rate, and sometimes even complete program or even driver lockups, it's most commonly the first two which happen in practice.

Does it only affect OpenGL or also DirectX? Are any particular API calls needed to trigger the bug? Are the lockups causing actual TDR resets or just hanging in the user-mode driver?

> I'm still unable to completely pinpoint how and where this bug actually occurs, but with lots of experiments I do know how to work around this issue. While this issue has only been reproducible on Intel UHD 620 and UHD 630 GPUs, HD 620 and HD 630 might also be affected, but I was unable to find anyone who has these, so I was unable send anyone my test program.

Can you post your test program here with details on expected behavior and behavior when encountering the bug?

**Comment by SonoSooS** [comment]
I'm sorry to have unwillingly wasted everyone's time with this.

I performed *a lot* of tests during multiple weeks, multiple operating systems, and multiple driver versions, and the short conclusions are "it works on my machine", and "either downgrade or upgrade both the driver and the OS".

I was so blinded by trying to resolve this issue, that I forgot to consider the affected users' software setup, namely their OS version and driver version, which only happens in some odd OS and driver combinations.


> Can you describe in more detail what the bug actually is and how to determine if it's affecting a game? If you have links to the discussion of this issue elsewhere, that would be helpful for context.

Depending on the OS and driver version, different things happen:
- old Windows (8.1) and old driver version, and macOS experience extremely unstable frametimes and stutters, and in certain cases even degraded performance, which can be mitigated with a simple glFinish after SwapBuffers
- somewhat old Windows (tested 1607 and 1703) with any functional driver: extreme stutters, dwm hangs/crashes, and occasional driver crashes
- Windows 10 1803 with Intel website drivers: extremely degraded performance, and occasional sutters, workarounds don't seem to affect this no matter what
- Windows 10 1803 with Windows Update drivers: the driver is not affected by any of the issues, and the workarounds especially hurt the performance (almost ~40-50% performance loss!)

Other combinations also exist, but I have run out of energy to double-check those, so I won't include those as to prevent possible false information.


There is [ppy/osu-framework#5164] and [ppy/osu-framework#5165] documenting this madness, but after this much research and testing, it seems like a rabbit hole too deep to be explored, and just better left alone, and the user has to solve it themselves (that is, by downgrading/upgrading Windows, or switching to Linux with Mesa drivers).

On macOS the user does not have any driver choices, so this workaround would be better suited for macOS instead, as opopsed to the original bug report, where it would've been applied to Windows. But even considering that this bug only happens in very specific use case where the GPU is overloaded with a very specific workload related to the memory bus being saturated due to extreme cache miss (see linked issues), so it's not a great idea to add this workaround either.

> Are any particular API calls needed to trigger the bug? Are the lockups causing actual TDR resets or just hanging in the user-mode driver?

I can't narrow it down to any specific API call, as it's quite arbitary when it happens.

One thing is certain: it happens if the GPU has a lot of L3 cache misses, and it stalls due to fetching an enormous amount of data from the memory bus, which confuses the affected drivers in different ways.

When the conditions are right for this bug triggering, most of the times it's just dwm crashing, but sometimes entire display driver crashes and resets as well.

> Does it occur when running OpenGL applications on top of ANGLE?

It still does happen, although it does mitigate the issue somewhat.

> Can you post your test program here with details on expected behavior and behavior when encountering the bug?

It turns out I had a typo in my test programs, which I managed to copypaste into all of them.
After noticing the typo, my test programs are no longer able to reproduce the issue, as now it's putting a "proper" load on the driver, as opposed to an irregular load as it did with the typo.
